server/src/repositories/job.repository.ts

```diff
@@ -249,24 +251,53 @@ export class JobRepository implements IJobRepository {
   }
 
   private getJobOptions(item: JobItem): JobsOptions | null {
+    let opts = { };
     switch (item.name) {
+      case JobName.NOTIFY_ALBUM_UPDATE:
       case JobName.STORAGE_TEMPLATE_MIGRATION_SINGLE: {
-        return { jobId: item.data.id };
+        opts = { jobId: item.data.id };
+        break;
       }
       case JobName.GENERATE_PERSON_THUMBNAIL: {
-        return { priority: 1 };
+        opts = { priority: 1 };
+        break;
       }
       case JobName.QUEUE_FACIAL_RECOGNITION: {
-        return { jobId: JobName.QUEUE_FACIAL_RECOGNITION };
-      }
-
-      default: {
-        return null;
+        opts = { jobId: JobName.QUEUE_FACIAL_RECOGNITION };
+        break;
       }
     }
+    if ((item?.data as IBaseJob)?.delay) {
+      (opts as any).delay = (item?.data as IBaseJob)?.delay;
+    }
+    return opts;
   }
 
   private getQueue(queue: QueueName): Queue {
     return this.moduleReference.get<Queue>(getQueueToken(queue), { strict: false });
   }
+
+  public async replaceQueuedOrInsertJob(item: JobItem, dataReplacer: DataTransformer): Promise<void> {
```

Comment:
You don't need to do any of this. Queueing a job with the same ID as another job will replace it.

Reply:
I'm not sure how to get that behavior, when I was testing earlier with `IJobRepository#queue` (which calls `#queueAll`) with the same job IDs, it silently drops the job until the first queued job is finished running. I have to get the job from bullmq, then remove it before enqueueing a new one.

Reply:
Oh darn, I thought it replaced the existing job, but it apparently just drops the new one. Removing and re-adding works too I guess. You can simplify it a bit by fetching by job ID instead of getting all the jobs and filtering.

Reply:
I think it'd be better to have a simple `remove` method to remove by a job ID, then maybe an `upsert` method that internally calls `remove` and `queue`.

Reply:
Actually, jobs also have a `changeDelay` method that does what we need in this case.

Reply:
I also have to change the data, the logic essentially does this:
* if an existing job to notify an album is found,
* remove it,
* check the recipients from the previous job, and add more people to notify if needed.

Say person p1 adds an image, and person p2 adds another image 20 seconds later.
If we solely remove the job and replace it, person p2 will never get a notification, as it was replaced with their notification to p1 that they added an image.

Reply:
Hmm, maybe it'd be better to make this a bit lazier. What if you scheduled a delayed job that, when, run, checks the `updatedAt` field of the album? If it's within a certain duration, it can re-queue itself and check again later. You don't need to worry about removing and re-adding this job; you just need to make sure it's queued.

Reply:
I'm imaging something like this:
* schedule only album ID and timestamp of first update.
* if job is still scheduled, push it back another few minutes but keep time of first update,
* when job fires, get all images added between timestamp on the job and now,
  * and notify those users.

does that sound good? also cleaner since there is no need to keep the users to notify as state, can just calculate that near the instant the notification is sent

Reply:
The first and third bullet points sound good to me.

For the second, I think it'd be better to always try to queue the job and check within the job if it should re-queue itself. The advantages are that it involves fewer requests and is more parallel-safe: when you check if the job exists and create it or change the delay based on that, you're doing 2 requests and conditions can change between those two.

If you queue the job and move on, it's one request with no race condition, taking advantage of the fact that later jobs with the same job ID are dropped (i.e. the timestamp of the first update is preserved).

Reply:
I think I might have something sorta stubbed out that might work, but I've run into a blocker:

`immich=# SELECT * FROM albums_assets_assets;`
               albumsId               |               assetsId
--------------------------------------|--------------------------------------
 3c8f611f-4d15-4ec6-89bf-343fb3b30156 | 3619bdcf-d8f3-41ff-a5fe-8750ff440e34
 3c8f611f-4d15-4ec6-89bf-343fb3b30156 | d0d3f9e1-24be-49e8-b12d-890d914bee4c

There doesn't seem to be anything that tracks when an image was added to an album, since I'm just passing when the album was updated, I need to get all new images. Would it make sense to add a `createdOn` to this table here? It doesn't look straightforward to accomplish with [typeorm](https://orkhan.gitbook.io/typeorm/docs/many-to-many-relations)

Reply:
Alright, I removed the weird DataReplacer function param and created a `remove` function on the JobRepository that just returns the job data if it's available to build out the new notification recipient list. Much easier to write a unit test for this case too.